core/lib/spree/testing_support/factories/taxon_factory.rb

```diff
@@ -9,12 +9,19 @@
 
 FactoryBot.define do
   factory :taxon, class: 'Spree::Taxon' do
+    transient do
+      parent { nil }
+      taxonomy { parent&.taxonomy || create(:taxonomy) }
+    end
+
     name { 'Ruby on Rails' }
-    taxonomy
-    parent_id { nil }
+    taxonomy_id { taxonomy.id }
+    parent_id { parent&.id || taxonomy.root.id }
 
     trait :with_icon do
-      icon { Spree::Core::Engine.root.join('lib', 'spree', 'testing_support', 'fixtures', 'blank.jpg').open }
+      after(:create) do |taxon|
```

Comment:
Note, `after(:build)` and setting the icon manually does not work 🤔  This change could break tests in stores if they were building taxons with icons?

Reply:
Sorry I mean, I had to change it to `after(:create)` for icons to be set correctly for nested taxons. If you checkout to this branch and revert this local change you can see that `core/spec/models/spree/taxon_spec.rb` fails because of this.

Reply:
I don't think this could be breaking for existing stores as long as they just use the `with_icon` trait. By the way, were you able to identify the cause of the spec failure without this change?

Reply:
Hmmm, it could if they are building taxons with icon, because the icon won't be set, whereas it was before. I haven't been able to identify the cause yet. 🤔

Reply:
Sorry, not sure to understand. Can you help with a spec in an existing store that was passing before this change and would fail if we merge?

Reply:
Hard to think of a plausible reason stores are using built taxons with icons and testing the icons behavior, so it's probably not of concern. But as a demonstration:

```ruby
  context "a built taxon's icon" do
    it "can be removed" do
      taxon = build(:taxon, :with_icon)

      # because the icon is no longer set
      expect(taxon.destroy_attachment(:icon)).to be_truthy
    end
  end
```

Reply:
But why do you say it's no longer set? I think it is in the `after(:create)` block, isn't it?

Reply:
No, because `after(:create)` is not triggered for build records @kennyadsl 
https://thoughtbot.com/blog/aint-no-calla-back-bot

Reply:
Ahhhhh! I wasn't thinking about the build vs create, but now I got it and I think you are right. 

Yes, that's not optimal. I think it's worth some more detailed investigation for the root cause of the error. I'm available to help with that if needed.

Reply:
@kennyadsl 
The old code for setting the icon no longer works as it sets the icon before building, and an `after_save` callback that comes from `awesome_nested_set` causes the icon to be lost.

[after_save callback](https://github.com/collectiveidea/awesome_nested_set/blob/6409bab837c9fa786509616bc409422b47ddbece/lib/awesome_nested_set/awesome_nested_set.rb#L69) > [move_to_new_parent](https://github.com/collectiveidea/awesome_nested_set/blob/6409bab837c9fa786509616bc409422b47ddbece/lib/awesome_nested_set/model/movable.rb#L122) > [move_to_child_of](https://github.com/collectiveidea/awesome_nested_set/blob/6409bab837c9fa786509616bc409422b47ddbece/lib/awesome_nested_set/model/movable.rb#L38) > [move_to](https://github.com/collectiveidea/awesome_nested_set/blob/6409bab837c9fa786509616bc409422b47ddbece/lib/awesome_nested_set/model/movable.rb#L99) > [reload_nested_set](https://github.com/collectiveidea/awesome_nested_set/blob/6409bab837c9fa786509616bc409422b47ddbece/lib/awesome_nested_set/model.rb#L263) (the cause!)

---

Doesn't make much sense to have save the icon on build records imo, seeing as in admin when you go to create a new taxon, it creates a dummy one and only then can you add an icon.

Reply:
Sorry for taking a lot of time to get back here. Yes, I think it's an acceptable compromise, and this sounds more like a bug for which we can introduce a breaking change.